test/storage/index_builder_test.cpp

```diff
@@ -0,0 +1,142 @@
+  const uint32_t num_threads_ = 4;
+
+  std::unique_ptr<DBMain> db_main_;
+  common::ManagedPointer<transaction::TransactionManager> txn_manager_;
+
+  // SqlTable
+  storage::SqlTable *sql_table_;
+  storage::ProjectedRowInitializer tuple_initializer_ =
+      storage::ProjectedRowInitializer::Create(std::vector<uint16_t>{1}, std::vector<uint16_t>{1});
+
+  common::WorkerPool thread_pool_{num_threads_, {}};
+
+ protected:
+  void SetUp() override {
+    thread_pool_.Startup();
+    db_main_ = terrier::DBMain::Builder().SetUseGC(true).SetUseGCThread(true).SetRecordBufferSegmentSize(1e6).Build();
+    txn_manager_ = db_main_->GetTransactionLayer()->GetTransactionManager();
+
+    auto idxcol = catalog::Schema::Column(
+        "attribute", type::TypeId::INTEGER, false,
+        parser::ConstantValueExpression(type::TransientValueFactory::GetNull(type::TypeId::INTEGER)));
+    auto nonidxcol = catalog::Schema::Column(
+        "attribute2", type::TypeId::INTEGER, false,
+        parser::ConstantValueExpression(type::TransientValueFactory::GetNull(type::TypeId::INTEGER)));
+    StorageTestUtil::ForceOid(&(idxcol), catalog::col_oid_t(1));
+    StorageTestUtil::ForceOid(&(nonidxcol), catalog::col_oid_t(2));
+    table_schema_ = catalog::Schema({idxcol, nonidxcol});
+    sql_table_ = new storage::SqlTable(db_main_->GetStorageLayer()->GetBlockStore(), table_schema_);
+    tuple_initializer_ = sql_table_->InitializerForProjectedRow({catalog::col_oid_t(1)});
+
+    std::vector<catalog::IndexSchema::Column> keycols;
+    keycols.emplace_back("", type::TypeId::INTEGER, false,
+                         parser::ColumnValueExpression(CatalogTestUtil::TEST_DB_OID, CatalogTestUtil::TEST_TABLE_OID,
+                                                       catalog::col_oid_t(1)));
+    StorageTestUtil::ForceOid(&(keycols[0]), catalog::indexkeycol_oid_t(1));
+    index_schema_ = catalog::IndexSchema(keycols, storage::index::IndexType::BWTREE, false, false, false, true);
+  }
+  void TearDown() override {
+    thread_pool_.Shutdown();
+
+    db_main_->GetTransactionLayer()->GetDeferredActionManager()->RegisterDeferredAction([=]() { delete sql_table_; });
+  }
+};
+
+/**
+ * This test creates multiple worker threads that all try to insert [0,num_inserts) as tuples in the table and into the
+ * primary key index. At completion of the workload, only num_inserts_ txns should have committed with visible versions
+ * in the index and table.
+ */
+// NOLINTNEXTLINE
+TEST_F(IndexBuilderTests, NullTable) {
+
+  auto index = (IndexBuilder().SetKeySchema(index_schema_).Build());
+
+  auto txn = txn_manager_->BeginTransaction();
+  EXPECT_NE(index, nullptr);
+
+  std::vector<TupleSlot> values;
+  index->ScanAscending(*txn, storage::index::ScanType::OpenBoth, 1, nullptr, nullptr, 0, &values);
+  EXPECT_EQ(values.size(), 0);
+  txn_manager_->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+}
+
+// NOLINTNEXTLINE
+TEST_F(IndexBuilderTests, OneTxnFullTable) {
```

Comment:
what about add more tests, with more txns, tables, and operations

Reply:
IndexBuilder is too low-level to care about other transactions. We have JDBC tests to verify transactional semantics